common/src/main/java/edu/kit/kastel/mcse/ardoco/core/common/util/SimilarityUtils.java

```diff
@@ -252,19 +263,137 @@ private static boolean checkRecommendedInstanceForSelection(ModelInstance instan
+                                        .size(), recommendedInstanceNameList.size()) >= similarity) {
             return true;
         }
         for (var nounMapping : ri.getNameMappings()) {
             for (var surfaceForm : nounMapping.getSurfaceForms()) {
                 var splitSurfaceForm = CommonUtilities.splitCases(surfaceForm);
                 var surfaceFormWords = CommonUtilities.splitAtSeparators(splitSurfaceForm);
                 if (SimilarityUtils.areWordsOfListsSimilar(instanceNames, surfaceFormWords, similarity) || SimilarityUtils.areWordsOfListsSimilar(
-                        longestNameSplit, surfaceFormWords, similarity)) {
+                        longestNameSplit, surfaceFormWords, similarity) || 1.0 * similarEntriesOfList(instanceNames, surfaceFormWords) / Math.max(instanceNames
+                                .size(), surfaceFormWords.size()) >= similarity || 1.0 * similarEntriesOfList(longestNameSplit, surfaceFormWords) / Math.max(
+                                        longestNameSplit.size(), surfaceFormWords.size()) >= similarity) {
                     return true;
                 }
             }
         }
         return false;
     }
+
+    //TODO: Move to common utilities
+    public static int similarEntriesOfList(ImmutableList<String> list1, ImmutableList<String> list2) {
+
+        MutableList<String> removed = Lists.mutable.empty();
+
+        for (var element : list1) {
+            if (list2.contains(element)) {
+                removed.add(element);
+            } else {
+                if (list2.select(e -> !removed.contains(e) && (e.contains(element) || element.contains(e))).size() == 1) {
+                    removed.add(element);
+                }
+            }
+        }
+
+        return removed.size();
+    }
+
+    private static boolean coversOtherPhraseVector(PhraseMapping phraseMapping1, PhraseMapping phraseMapping2) {
+
+        MutableMap<Word, Integer> phraseVector1 = phraseMapping1.getPhraseVector().toMap();
+        MutableMap<Word, Integer> phraseVector2 = phraseMapping2.getPhraseVector().toMap();
+
+        return phraseVector1.keySet().containsAll(phraseVector2.keySet());
+    }
+
+    private static boolean containsAllNounMappingsOfPhraseMapping(TextState textState, PhraseMapping phraseMapping1, PhraseMapping phraseMapping2) {
+        return phraseMapping1.getNounMappings(textState).containsAllIterable(phraseMapping2.getNounMappings(textState));
+    }
+
+    private static double cosineSimilarity(Map<Word, Integer> phraseVector1, Map<Word, Integer> phraseVector2) {
+
+        CosineSimilarity cosineSimilarity = new CosineSimilarity();
+
+        Map<CharSequence, Integer> vector1 = phraseVector1.entrySet().stream().collect(Collectors.toMap(e -> e.getKey().getText(), Map.Entry::getValue));
+        Map<CharSequence, Integer> vector2 = phraseVector2.entrySet().stream().collect(Collectors.toMap(e -> e.getKey().getText(), Map.Entry::getValue));
```

Comment:
Typically you would use something like "firstVector" and "secondVector" to be sure that they are not accidentally swapped